### PR TITLE
remove deprecated old type+tag filters and allow groupById and aggregatebyGeometry after mapping/filtering

### LINKED ISSUE #324: [oshdb-api] drop deprecated mapreducer methods
* [ ] old filters :arrow_right: #451 
  * remove `osmType`
  * remove `osmTag`
  * update tests (merge `TestOSMDataFilters` with `TestOhsomeFilter`, replace deprecated methods with corresponding _filters_ where it makes sense)
* [ ] `forEach`
* …?

### PR DESCRIPTION
**Pull request by tyrasd**
addresses #324:

* removes old `osmType` filter methods
* removes old `osmTag` filter methods
* gets rid of (`osmTag(key, valueRegexp)`) method. reason: it was rarely used and never really worked when using an ignite backend. Also, it can be implemented relatively easily manually if needed in an analysis.
* implements `groupById` and `aggregateByGeometry` also after a filter, map or flatMap function was specified on a mapReducer. (solving the issue outlined in https://github.com/GIScience/oshdb/issues/442#issue-1204238985)
  the following corner case still causes an `UnsupportedOperationException`: `mapRed.filter("…").aggregateBy(…).aggregateByGeometry(…)` but it can be worked around relatively easily by either swapping the two aggregateBy's (`mapRed.filter("…").aggregateByGeometry(…).aggregateBy(…)`) or by delaying the filtering (`mapRed.aggregateBy(…).aggregateByGeometry(…).filter("…")`).
 
### Corresponding issue
Closes #324

### Checklist
- [x] My code follows the [code-style](https://github.com/GIScience/oshdb/blob/master/CONTRIBUTING.md) rules, and I have checked on the [static analyses](https://jenkins.ohsome.org/job/oshdb/view/change-requests/) and [benchmark](https://reports.ohsome.org/oshdb-benchmarks/) (if applicable) results
- [x] I have commented my code
- [x] I have written javadoc (required for public classes and methods)
- [x] I have added sufficient unit tests
- ~~I have made corresponding changes to the [documentation](https://github.com/GIScience/oshdb/tree/master/documentation)~~
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/oshdb/blob/master/CHANGELOG.md)
- ~~I have adjusted the [examples](https://gitlab.gistools.geog.uni-heidelberg.de/giscience/big-data/ohsome/oshdb-examples) or [created an issue](https://gitlab.gistools.geog.uni-heidelberg.de/giscience/big-data/ohsome/oshdb-examples/-/issues/new) in the corresponding repository~~
- ~~[ ] I have adjusted the [benchmark](https://reports.ohsome.org/oshdb-benchmarks/) or [created an issue](https://gitlab.gistools.geog.uni-heidelberg.de/giscience/big-data/ohsome/oshdb-benchmarks/-/issues/new) in the corresponding repository~~

### CODE REVIEW

**Comment by tyrasd** [comment]
I'm not sure why this apparently breaks the following two tests:

```
TestMapReduceOSHDBH2Singlethread.testTimeoutMapReduce Expected exception: org.heigit.ohsome.oshdb.util.exceptions.OSHDBTimeoutException
TestMapReduceOSHDBH2Singlethread.testTimeoutStream Expected exception:org.heigit.ohsome.oshdb.util.exceptions.OSHDBTimeoutException
```

It works fine locally here. Will need to investigate :thinking:

**Comment by rtroilo** [comment]
> I'm not sure why this seemingly breaks the following two tests on Jenkins:
> 
> ```
> TestMapReduceOSHDBH2Singlethread.testTimeoutMapReduce Expected exception: org.heigit.ohsome.oshdb.util.exceptions.OSHDBTimeoutException
> TestMapReduceOSHDBH2Singlethread.testTimeoutStream Expected exception:org.heigit.ohsome.oshdb.util.exceptions.OSHDBTimeoutException
> ```
> 
> These two tests work fine locally here. Will need to investigate thinking

Don't worry about this tests, we fixed them already in #450

**Comment by joker234** [comment]
> I'm not sure why this seemingly breaks the following two tests on Jenkins:
> 
> ```
> TestMapReduceOSHDBH2Singlethread.testTimeoutMapReduce Expected exception: org.heigit.ohsome.oshdb.util.exceptions.OSHDBTimeoutException
> TestMapReduceOSHDBH2Singlethread.testTimeoutStream Expected exception:org.heigit.ohsome.oshdb.util.exceptions.OSHDBTimeoutException
> ```
> 
> These two tests work fine locally here. Will need to investigate thinking

Even though we fixed a lot in #450, the tests were still failing. I suspect that `.filter(…)` is much faster than `.osmEntityFilter(…)`. The tests worked (locally) sometimes for me with a lower timeout setting (like 5 or 10 ms), sometimes not. With setting the timeouts to 1 ms it worked all the time (see 4704607).

File: oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java [hunk 1]
```diff
@@ -866,19 +597,42 @@ public MapReducer<X> filter(String f) {
...
-    @SuppressWarnings("unchecked") // now in the reduce step the backend will return a list of items
-    MapReducer<List<X>> result = (MapReducer<List<X>>) ret;
-    return result;
+    if (!this.mappers.isEmpty()) {
+      // for convenience, we allow one to set this function even after some map functions were set.
+      // if some map / flatMap functions were already set:
+      // "rewind" them first, apply the grouping and then re-apply the map/flatMap functions
+      // accordingly
+      MapReducer<X> ret = this.copy();
+      List<MapFunction> mappers = new LinkedList<>(ret.mappers);
```
Comment (rtroilo) [inline oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java:609]:
```suggestion
      List<MapFunction> mappers = new ArrayList<>(ret.mappers);
```

Comment (rtroilo) [reply oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java:609]: Also rename `mappers` to something else as this variable hides the class variable `mappers` and this is reported as a code smell. The same is true in `aggregateByGeometry`.

File: oshdb-api/src/test/java/org/heigit/ohsome/oshdb/api/tests/TestMapAggregateByGeometry.java [hunk 2]
```diff
@@ -139,16 +136,16 @@ void testZerofill() throws Exception {
 
   @Test
   void testCombinedWithAggregateByTimestamp() throws Exception {
-    SortedMap<OSHDBCombinedIndex<OSHDBTimestamp, String>, Integer> result =
+    SortedMap<OSHDBCombinedIndex<String, OSHDBTimestamp>, Integer> result =
         createMapReducerOSMEntitySnapshot()
             .timestamps(timestamps1)
-            .aggregateByTimestamp()
             .aggregateByGeometry(getSubRegions())
+            .aggregateByTimestamp(OSMEntitySnapshot::getTimestamp)
```
Comment (rtroilo) [inline oshdb-api/src/test/java/org/heigit/ohsome/oshdb/api/tests/TestMapAggregateByGeometry.java:143]: What was the reason to change the order of the aggregation from `timestamp, geometry` to `geometry, timestamp`.

File: oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java (hunk 1 above)
Comment (joker234) [reply oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java]: Found it in the aforementioned methods and `aggregateByTimestamp`, see 209ae2c3.

Comment (tyrasd) [reply oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java]: fixed in 40dba317 by simplifying the implementation of the "automatic" aggregation methods. As as sideeffect, now all(?) aggregation combinations are allowed and possible regardless of their order, for example: `mapRed.aggregateByGeometry(…).aggregateByTimestamp()` or `mapRed.map(…).aggregateByTimestamp().aggregateByGeometry(…)`

File: oshdb-api/src/test/java/org/heigit/ohsome/oshdb/api/tests/TestMapAggregateByGeometry.java (hunk 2 above)
Comment (tyrasd) [reply oshdb-api/src/test/java/org/heigit/ohsome/oshdb/api/tests/TestMapAggregateByGeometry.java]: fixed in 40dba317 by simplifying the implementation of the "automatic" aggregation methods. As as sideeffect, now all(?) aggregation combinations are allowed and possible regardless of their order, for example: `mapRed.aggregateByGeometry(…).aggregateByTimestamp()` or `mapRed.map(…).aggregateByTimestamp().aggregateByGeometry(…)`

**Review by rtroilo — CHANGES_REQUESTED** [review]
just a quick comment, why the widening to `object`, see suggestions, is necessary. It seems for me that the tests run successfully through even without this.

**Review by rtroilo — DISMISSED** [review]

**Review by rtroilo — APPROVED** [review]
Thank you .

**Pull request merged**